proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8847,24 +8847,34 @@ lemma bound_head_time_buffer:
   done
 
 lemma head_time_buffer_true_imp_buffer:
-  "\<lbrakk>the (head_time_buffer usage s); pred_map \<top> (sc_refill_cfgs_of s) (cur_sc s)\<rbrakk>
-   \<Longrightarrow> pred_map (\<lambda>cfg. max_time - r_time (hd ((scrc_refills cfg))) \<ge> 5 * MAX_PERIOD)
-                (sc_refill_cfgs_of s) (cur_sc s)"
-  apply (frule bound_head_time_buffer[where usage=usage])
-  apply (clarsimp simp: head_time_buffer_def obind_def read_sched_context_def ogets_def
-                        vs_all_heap_simps
-                 split: kernel_object.splits)
+  "pred_map \<top> (sc_refill_cfgs_of s) (cur_sc s)
+   \<Longrightarrow> the (head_time_buffer usage s)
+       = (pred_map (\<lambda>cfg. r_time (scrc_refill_hd cfg) < MAX_RELEASE_TIME)
+                   (sc_refill_cfgs_of s) (cur_sc s)
+          \<and> pred_map (\<lambda>cfg. r_amount (scrc_refill_hd cfg) \<le> usage )
+                     (sc_refill_cfgs_of s) (cur_sc s))"
+  apply (fastforce simp: head_time_buffer_def obind_def read_sched_context_def ogets_def
+                         vs_all_heap_simps MAX_RELEASE_TIME_def
+                  split: kernel_object.splits)
+  done
+
+lemma unat_MAX_RELEASE_TIME:
+  "unat MAX_RELEASE_TIME = unat max_time - 5 * unat MAX_PERIOD"
```

Comment:
nitpick: it wouldn't hurt to have this earlier, maybe in MachineOps as well if `MAX_RELEASE_TIME` is moved